logging/src/commonMain/kotlin/Pool.kt

```diff
@@ -0,0 +1,16 @@
+package com.juul.tuulbox.logging
+
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+
+/** While a [Pool] is logically thread safe, Kotlin/Native's memory model requires @ThreadLocal on instances of this. */
+internal class Pool<T>(
+    private val factory: () -> T,
+    private val refurbish: (T) -> Unit
+) {
+    private val lock = reentrantLock()
+    private val cache = ArrayDeque<T>()
```

Comment:
Does the queue (as apposed to an `atomic`) protect against someone logging from within a log lambda?
Otherwise, would an `atomic` have sufficed?

Reply:
Re. `ArrayDeque`, I just default to it instead of `List` when I'm doing stack/queue operations, since it's more explicit. For stack operations (like here) it's basically the same.

Re. `atomic`, it doesn't work for the high throughput, low-allocation goal of this pool. In `DispatchLogger`, for example, we're using `consumers.update { it + consumer }` -- which allocates a new list and copies all the plus the new one (I'm not worried about the O(N^2) inefficiency here, because N is super small and setup only happens once). We have to do that copy for two reasons.
1. `update` is loop based. It gets the current value, performs the update function, then does a `compareAndSet` to make sure nothing else changed the value. This means that the block can be called more than once, so it's not safe to mutate the value. But even if it wasn't a loop...
2. `atomic` is basically a pointer. While it's okay to change what it points at (from any thread), in Kotlin/Native the thing pointed at is still frozen. This would also prevent us from mutating a list.

The solution here, albeit ugly, should be fast on every platform. Effectively...
- On JVM: we're using one pool shared by all threads. Getting a value from the pool and returning a value to it holds the lock, but everything else can live outside it. Realistically, there should be very little contention and when there is, it should resolve itself quickly.
- On JS: the lock is implemented as a no-op.
- On Native: the `Pool` itself is `@ThreadLocal`, so the `lock`/`cache` are each per-thread. As such, while the lock exists, there will never be any contention and it will always happy-path.

Reply:
Thanks for the incredibly thorough explanation. Super clever approach with the `@ThreadLocal`. 👍

Reply:
And yet, somehow, despite writing a small essay, I managed to miss the original question.

It _should_ be safe to nest log calls, both inside a `Logger` (although it will need conditional logic, or else it will infinitely loop) and within a `Log` call such as `Log.verbose { Log.debug { "a" }; "b" }`. I'll admit I haven't tested either of these cases.